File: remove_invalid_dfs.py
# // Time Complexity :O(exponential)
# // Space Complexity :O(n)
# // Did this code successfully run on Leetcode :yes
# // Any problem you faced while coding this :giving time limisted exceeded when submitting, but when the same input was given as text case works fine


# // Your code here along with comments explaining your approach

class Solution:
    def __init__(self):
        self.result=[]
        self.hset=[]
        self.max=0
    def removeInvalidParentheses(self, s: str) -> List[str]:
    
        self.dfs(s)
        return self.result
    def dfs(self,s):
        #base
        if s in self.hset or len(s)<self.max:
            return
        #logic
        self.hset.append(s)
        if self.isvalid(s):
            if len(s)>self.max:
                self.result=[]
                self.max=len(s)
                self.result.append(s)
            else:
                self.result.append(s)

                
        else:
            for i in range(len(s)):
                subs=s[:i]+s[i+1:]
                self.dfs(subs)
            
        
    def isvalid(self,curr):
        count=0
        for i in curr:
            
            if i=='(':
                count+=1
            elif i==')':
                count-=1
            if count==-1:
                return False
        return count==0
            

File: RemoveInvalidDFS.java
class Solution {
    //tc- n^n sc-n^n
    HashSet<String> set;
    List<String> result;
    int max;
    public List<String> removeInvalidParentheses(String s) {
        result = new ArrayList<>();
        if(s == null || s.length() == 0) return null;
         set = new HashSet<>();
         dfs(s);
         return result;
    }

    private void dfs(String s)
    {
      //base case
      if(s.length() < max || set.contains(s)) return;
      if(isValid(s))
      {
          if(s.length() > max)
          {
              max = s.length();
              result = new ArrayList<>();
              result.add(s);
          }
          else if(s.length() == max)
          {
              result.add(s);
          }
          set.add(s);
      }

      //logic
      set.add(s);
      for(int i=0;i<s.length();i++)
      {
          char c = s.charAt(i);
          if(Character.isLetter(c)) continue;
          String child = s.substring(0,i) + s.substring(i+1);
          dfs(child);
      }
    }

    private boolean isValid(String s)
    {
        int count =0;
        for(int i=0;i<s.length();i++)
        {
            char c = s.charAt(i);
            if(Character.isLetter(c)) continue;
            if(c == '(') count++;
            else if(c == ')') count--;
            if(count <0) return false;
        }
        return count ==0;
    }
}